pypose/module/pf.py

```diff
@@ -0,0 +1,212 @@
+        ... print('Est error:', (states - estim).norm(dim=-1))
+        Est error: tensor([5.3627, 0.5640, 0.0953, 0.0447, 0.0936])
+
+    Note:
+        Implementation is based on Section 15.2 of this book
+
+        * Dan Simon, `Optimal State Estimation: Kalman, H∞, and Nonlinear Approaches
+          <https://onlinelibrary.wiley.com/doi/book/10.1002/0470045345>`_,
+          Cleveland State University, 2006
+    '''
+
+    def __init__(self, model, Q=None, R=None, particles=1000):
+        super().__init__(model, Q, R)
+        self.particles = particles
+
+    def forward(self, x, y, u, P, Q=None, R=None, t=None, n=None):
+        r'''
+        Performs one step estimation.
+
+        Args:
+            x (:obj:`Tensor`): estimated system state of previous step
+            y (:obj:`Tensor`): system observation at current step (measurement)
+            u (:obj:`Tensor`): system input at current step
+            P (:obj:`Tensor`): state estimation covariance of previous step
+            Q (:obj:`Tensor`, optional): covariance of system transition model
+            R (:obj:`Tensor`, optional): covariance of system observation model
+            n (:obj:`Tensor`, optional): covariance weight for randomly generate particles
+
+        Return:
+            list of :obj:`Tensor`: posteriori state and covariance estimation
+        '''
+        # Upper cases are matrices, lower cases are vectors
+
+        Q = Q if Q is not None else self.Q
+        R = R if R is not None else self.R
+        n = x.size(-1) if n is None else n
+        self.model.set_refpoint(state=x, input=u, t=t)
+
+        xp = self.generate_particles(x, n * P)
+        xs = self.model.state_transition(xp, u, t)
+        ye = self.model.observation(xs, u, t)
+        q = self.relative_likelihood(y, ye, R)
+        xr = self.resample_particles(q, xs)
+
+        x = xr.mean(dim=0)
+        ex = xr - x
+        weight = torch.tensor([1 / self.particles])
+        P = self.compute_cov(ex, ex, weight, Q)
+
+        return x, P
+
+    def generate_particles(self, x, P):
+        r'''
+        Randomly generate particles
+
+        Args:
+            x (:obj:`Tensor`): estimated system state of previous step
+            P (:obj:`Tensor`): state estimation covariance of previous step
+
+        Return:
+            list of :obj:`Tensor`: particles
+        '''
+        m = MultivariateNormal(x, P)
+        return m.sample(torch.Size([self.particles]))
+
+    def relative_likelihood(self, y, ye, R):
+        r'''
+        Compute the relative likelihood
+        '''
+        q = MultivariateNormal(ye, R).log_prob(y).exp()
+        return F.normalize(q, p=1, dim=-1)
+
+    def resample_particles(self, q, x):
+        r'''
+        Resample the set of a posteriori particles
+        '''
+        r = torch.rand(self.particles, dtype=x.dtype, device=x.device)
+        cumsumq = torch.cumsum(q, dim=0)
+        cumsumq[-1] = 1.0
```

Comment:
if the above works, then this setting to `1` will not needed.

Reply:
It is a pity that `MultivariateNormal` subclasses [`Distribution`](https://pytorch.org/docs/stable/distributions.html?highlight=distribution#torch.distributions.distribution.Distribution), but doesn't implement `cdf`, otherwise we can combine `relative_likelihood` and `resample_particles` into one function like this:
```python
def resample_relative_likelihood(self, y, ye, R, x):
    r = torch.rand(self.particles, dtype=x.dtype, device=x.device)
    cumsumq = MultivariateNormal(ye, R).cdf(y)
    return x[torch.searchsorted(cumsumq, r)]
```

Reply:
Sorry to see it so late, thanks for your advice.